src/libsyntax/ext/expand.rs

```diff
@@ -1103,6 +1156,40 @@ pub fn expand_crate(mut cx: ExtCtxt,
     return (ret, cx.syntax_env.names);
 }
 
+// Expands crate using supplied MacroExpander - allows for
+// non-standard expansion behaviour (eg. stepwise)
+pub fn expand_crate_with_expander(expander: &mut MacroExpander,
+                                  user_exts: Vec<NamedSyntaxExtension>,
+                                  mut c: Crate) -> (Crate, HashSet<Name>) {
+    if std_inject::no_core(&c) {
+        expander.cx.crate_root = None;
+    } else if std_inject::no_std(&c) {
+        expander.cx.crate_root = Some("core");
+    } else {
+        expander.cx.crate_root = Some("std");
+    }
```

Comment:
Is this logic copied from somewhere? If so, could it be factored out into a function?